kedro/io/partitioned_data_set.py

```diff
@@ -58,6 +59,18 @@
 S3_PROTOCOLS = ("s3", "s3a", "s3n")
 
 
+def _grandparent(path: str) -> str:
+    path_obj = PurePosixPath(path)
+    grandparent = path_obj.parents[1]
+    if grandparent.name != path_obj.name:
+        last_three_parts = path_obj.relative_to(*path_obj.parts[:-3])
+        raise DataSetError(
+            f"`{path}` is not a well-formed versioned path ending with "
+            f"`filename/timestamp/filename` (got `{last_three_parts}`)."
+        )
+    return str(grandparent)
```

Comment:
Maybe?
```suggestion
    return grandparent.as_posix()
```

Reply:
I couldn't remember whether this line had any impact on S3 path handling, so I tried running through an example real quick:

```
>>> from pathlib import PurePosixPath
>>> path_obj = PurePosixPath("s3://bucket-name/path/to/folder/new/partition.csv")
>>> path_obj.parents[1]
PurePosixPath('s3:/bucket-name/path/to/folder')
```

Since `path_obj` is a `PurePosixPath`, `grandparent` is, too. As a result, I don't think this does anything (the `as_posix` call basically adds a path separator substitution that won't match anything).

Still, if this makes it more clear, I'll change it; let me leave it for last.

Reply:
Cool, `str(grandparent)` makes sense then 👍